### PR TITLE
[api] Refactor the attribute controller a little

### PR DESCRIPTION
**Pull request by coolo**
Extract functions and make the routes match the request

### CODE REVIEW

File: src/api/app/controllers/attribute_controller.rb [hunk 1]
```diff
@@ -34,59 +35,60 @@ def index
...
-      return
+  def show_namespace_definition
+    @an = AttribNamespace.where(name: ensure_namespace).select(:id, :name).first
+    unless @an
+      render_error message: "Unknown attribute namespace '#{namespace}'",
+        status: 404, errorcode: 'unknown_attribute_namespace'
     end
+  end
 
+  def require_admin
```
Comment (Ana06) [inline src/api/app/controllers/attribute_controller.rb]: why is this not private? :thinking:

File: src/api/app/controllers/attribute_controller.rb [hunk 2]
```diff
@@ -34,59 +35,60 @@ def index
...
-    if request.get?
-      @an = AttribNamespace.where(name: namespace).select(:id, :name).first
-      unless @an
-        render_error message: "Unknown attribute namespace '#{namespace}'",
-          status: 404, errorcode: 'unknown_attribute_namespace'
-      end
-      return
+  def show_namespace_definition
+    @an = AttribNamespace.where(name: ensure_namespace).select(:id, :name).first
+    unless @an
```
Comment (Ana06) [inline src/api/app/controllers/attribute_controller.rb]: Rubocop won't like this :disappointed:

File: src/api/app/controllers/attribute_controller.rb [hunk 3]
```diff
@@ -34,59 +35,60 @@ def index
...
+      render_error message: "Unknown attribute namespace '#{namespace}'",
+        status: 404, errorcode: 'unknown_attribute_namespace'
     end
+  end
 
+  def require_admin
     # namespace definitions must be managed by the admin
     return unless extract_user
     unless User.current.is_admin?
```
Comment (Ana06) [inline src/api/app/controllers/attribute_controller.rb]: rubocop should complain about this as well :thinking:

File: src/api/app/controllers/attribute_controller.rb (hunk 1 above)
Comment (bgeuken) [inline src/api/app/controllers/attribute_controller.rb]: This is already defined [here](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/controllers/application_controller.rb#L33)

File: src/api/app/controllers/attribute_controller.rb [hunk 4]
```diff
@@ -302,4 +271,25 @@ def find_attribute_container
     params[:namespace] = name_parts[0]
     params[:name] = name_parts[1]
   end
+
+  private
+
+  def ensure_namespace
```
Comment (Ana06) [inline src/api/app/controllers/attribute_controller.rb]: who is this related with rubocop? It seems like it is in the wrong commit

File: src/api/app/controllers/attribute_controller.rb [hunk 5]
```diff
@@ -92,30 +75,17 @@ def update_namespace_definition
...
     end
     name = params[:name]
 
     @at = @ans.attrib_types.find_by(name: name)
-    unless @at
-        render_error message: "Unknown attribute '#{namespace}':'#{name}'",
-          status: 404, errorcode: 'unknown_attribute'
-      end
-    end
+    return if @at
```
Comment (Ana06) [inline src/api/app/controllers/attribute_controller.rb]: add it to the commit where the offenses where introduced.... :unamused:

File: src/api/app/controllers/attribute_controller.rb [hunk 6]
```diff
@@ -87,54 +87,52 @@ def show_attribute_definition
...
-  def attribute_definition
+  def delete_attribute_definition
     # permission check via User model
     return unless extract_user
 
-    if request.post? || request.put?
-      logger.debug '--- updating attribute type definitions ---'
-
-      xml_element = Xmlhash.parse(request.raw_post)
+    if at = @ans.attrib_types.where('name = ?', name).first
```
Comment (Ana06) [inline src/api/app/controllers/attribute_controller.rb]: `@ans.attrib_types.where('name = ?', name).first` can be written like:

`@ans.attrib_types.where(name: name).first` or even better:

`@ans.attrib_types.find_by(name: name)`

File: src/api/app/controllers/attribute_controller.rb [hunk 7]
```diff
@@ -87,54 +87,52 @@ def show_attribute_definition
...
+        message: "Illegal request: PUT/POST #{request.path}: path does not match content"
+      return
+    end
 
-      logger.debug '--- finished updating attribute namespace definitions ---'
-      #--- end update attribute namespace definitions ---#
-      render_ok
-    elsif request.delete?
-      at = @ans.attrib_types.where('name = ?', name).first
+    entry = @ans.attrib_types.where('name = ?', name).first
```
Comment (Ana06) [inline src/api/app/controllers/attribute_controller.rb]: you can write it like `@ans.attrib_types.find_by(name: name)` and this code is repeated... :disappointed:

File: src/api/app/controllers/attribute_controller.rb [hunk 8]
```diff
@@ -87,54 +87,52 @@ def show_attribute_definition
...
-        message: "Illegal request: POST #{request.path}"
+      entry = AttribType.new(name: name, attrib_namespace: @ans)
+      authorize entry, :create?
+
+      logger.debug '* create new attribute definition'
+      entry.update_from_xml(xml_element)
     end
+
+    logger.debug '--- finished updating attribute namespace definitions ---'
+    #--- end update attribute namespace definitions ---#
```
Comment (Ana06) [inline src/api/app/controllers/attribute_controller.rb]: is this really needed? :unamused:

File: src/api/app/controllers/attribute_controller.rb (hunk 1 above)
Comment (bgeuken) [reply src/api/app/controllers/attribute_controller.rb]: @coolo Ah. I didn't see that there was an admin check in some controller already.

In that leave it as is. Let's not change the API behaviour.

File: src/api/app/controllers/attribute_controller.rb (hunk 7 above)
Comment (coolo) [reply src/api/app/controllers/attribute_controller.rb]: I know - I just moved lines

File: src/api/app/controllers/attribute_controller.rb (hunk 1 above)
Comment (coolo) [reply src/api/app/controllers/attribute_controller.rb]: I reconsidered and I don't think anyone relies on the different error message for requiring admin.

File: src/api/app/controllers/attribute_controller.rb (hunk 4 above)
Comment (coolo) [reply src/api/app/controllers/attribute_controller.rb]: I squashed now everything in one commit

**Comment by coolo** [comment]
all tests pass, all feedback worked on

**Comment by bgeuken** [comment]
In future please try to create commits per logical change, eg. one commit for moving a method somewhere else and another for refactoring that method.
This makes it much easier to review the code and spot issues.

**Comment by coolo** [comment]
superseded by #5165
